### PR TITLE
Add hooks platform

### PR DESCRIPTION
**Pull request by chrisccoulson**
This adds a KeyData platform that permits sealing and recovery of keys
via hooks provided via a device's gadget and kernel snaps, for platforms
where we don't have a native secboot platform.

### CODE REVIEW

**Comment by chrisccoulson** [comment]
One thing I've just noticed here is that I've missed some integration tests between the new platform and `secboot.KeyData`. I'll push another update with those in, although it doesn't stop anyone reviewing what's already here.

File: hooks/platform.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/snapcore/secboot"
+)
+
+const (
+	platformName = "fde-hooks-v3"
```
Comment (pedronis) [inline hooks/platform.go:42]: I wonder if for consistency with the other naming and the context of all of this we need the "fde-" of the name here?

File: hooks/keydata.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		aead, err := cipher.NewGCM(b)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot create AEAD for AEAD compat: %w", err)
+		}
+		ciphertext = aead.Seal(nil, nonce, payload, aad)
+
+		var encryptedKey []byte
+		encryptedKey, handle, err = keyProtector.ProtectKey(rand, symKey, nil)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect symmetric key for AEAD compat: %w", err)
```
Comment (pedronis) [inline hooks/keydata.go:200]: it would be good to have a test for the error path coming from the Protector

File: hooks/keydata.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		}
+
+		aeadCompat = &aeadCompatData{
+			Nonce:        nonce,
+			EncryptedKey: encryptedKey,
+		}
+	default:
+		ciphertext, handle, err = keyProtector.ProtectKey(rand, payload, aad)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect key: %w", err)
```
Comment (pedronis) [inline hooks/keydata.go:210]: same here

File: hooks/keydata.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+// key that was originally supplied to [NewProtectedKey].
+//
+// On success, this will automatically update the corresponding *[secboot.KeyData]
+// that this key data was created from using [NewKeyData].
+func (d *KeyData) SetAuthorizedBootModes(key secboot.PrimaryKey, modes ...string) error {
+	if err := d.data.Scope.SetAuthorizedBootModes(key, d.k.Role(), modes...); err != nil {
+		return err
+	}
+	return d.k.MarshalAndUpdatePlatformHandle(d)
+}
```
Comment (pedronis) [inline hooks/keydata.go:284]: these methods don't seem to be exercised in tests

File: hooks/platform.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		}
+	}
+
+	keyRevealerMu.Lock()
+	defer keyRevealerMu.Unlock()
+
+	switch {
+	case kd.data.AEADCompat != nil:
+		symKey, err := keyRevealer.RevealKey(kd.data.Handle, kd.data.AEADCompat.EncryptedKey, nil)
+		if err != nil {
```
Comment (pedronis) [inline hooks/platform.go:82]: same for the revealer

File: hooks/platform.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			return nil, &secboot.PlatformHandlerError{
+				Type: secboot.PlatformHandlerErrorInvalidData,
+				Err:  fmt.Errorf("cannot recover key: %w", err),
+			}
+		}
+		return payload, nil
+	default:
+		payload, err := keyRevealer.RevealKey(kd.data.Handle, encryptedPayload, aad)
+		if err != nil {
```
Comment (pedronis) [inline hooks/platform.go:114]: and here

**Review by pedronis — COMMENTED** [review]
thank you, did a first pass

File: hooks/platform.go (hunk 1 above)
Comment (chrisccoulson) [reply hooks/platform.go:42]: I'm not sure - I just based this on what's currently in snapd (`fde-hooks-v2`)

Comment (pedronis) [reply hooks/platform.go:42]: ah, I forgot about that. This is not unreasonable then

File: hooks/keydata.go (hunk 2 above)
Comment (chrisccoulson) [reply hooks/keydata.go]: I've added a test for this now.

File: hooks/keydata.go (hunk 3 above)
Comment (chrisccoulson) [reply hooks/keydata.go]: And here

File: hooks/platform.go (hunk 5 above)
Comment (chrisccoulson) [reply hooks/platform.go:82]: I've added tests for key revealer errors, plus tests for several other error paths.

File: hooks/platform.go (hunk 6 above)
Comment (chrisccoulson) [reply hooks/platform.go:114]: There's a test for this one as well.

**Comment by chrisccoulson** [comment]
I've added most tests now with the exception of tests for `KeyData.SetAuthorizedSnapModels` and `KeyData.SetAuthorizedBootModes`,which I'll do after the weekend. I've also added a bunch of extra tests for error paths in `hooksPlatform.RecoverKeys`.

I'm going to add some tests for the JSON serialization next week as well. In order to avoid non-determinism in the outputs here (because of the ECDSA signatures in `bootscope` which consume randomness), I've added an `io.Reader` for functions that consume random bytes so that unit tests can override the source of randomness with fixed sets of bytes where they need determinism, and plumbed this into `bootscope`. I've had to add an additional helper to `internal/testutil` called `BypassMaybeReadByte` to bypass the usage of go's internal `randutil.MaybeReadByte` which is used in various go crypto packages in order to make the results of APIs non-deterministic even for the same sequence of input bytes.

File: hooks/keydata.go (hunk 4 above)
Comment (chrisccoulson) [reply hooks/keydata.go]: I've added some tests for both of these methods now.

**Comment by chrisccoulson** [comment]
The 1.22 tests are failing unfortunately due to change in nonce generation when creating ECDSA signatures, although it's still deterministic (well, it is with my `BypassMaybeReadByte` hack. I'm re-working them to try to keep the hardcoded byte sequences in the tests as I think it's important for detecting changes that might have backwards compatibility implications (particularly with the marshalling / unmarshalling tests - and I'd like to use this same approach for testing in other packages), but this is a spanner in the works I could have done without tbh :(

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**